### PR TITLE
Update build

### PR DESCRIPTION
**Pull request by bitraid**
## Summary
Some improvements to the build system and github workflow.

Added all the [environments of msys2](https://www.msys2.org/docs/environments/), except MSYS (doesn't target native Windows programs) and CLANGARM64 (cross compilation from x86 is not supported yet).

Also added cross compile support from linux using llvm-mingw, for all supported targets, including arm32/64 (tested in qemu).

Some other small improvements are included, each in its own commit for easier review.

I didn't increase the Far version, since the changes don't affect the actual program.
 
## References
https://github.com/FarGroup/FarManager/discussions/749#discussioncomment-8577527

## Checklist
- [x] I have followed the [contributing guidelines](https://github.com/FarGroup/FarManager/blob/master/CONTRIBUTING.md).
- [x] I have discussed this with project maintainers:  <br/>
If not checked, I accept that this work might be rejected in favor of a different grand plan.<br/>

### CODE REVIEW

File: far/main.cpp [hunk 1]
```diff
@@ -476,7 +476,7 @@ static void handle_exception(function_ref<bool()> const Handler)
 	throw;
 }
 
-#ifndef _WIN64
+#if !defined(_WIN64) && !defined(_ARM_)
```
Comment (alabuzhev) [inline far/main.cpp:479]: Might be simpler to `#ifdef _M_IX86` (same in other similar places).

File: plugins/luamacro/luasdk/ARM32/lua51.dll
Comment (alabuzhev) [inline plugins/luamacro/luasdk/ARM32/lua51.dll:1]: I see that lua5.1.dll and lpeg.dll depend on all those api-ms-win-crt-whatever, while the ARM64 dlls we already have in the repo don't.
Perhaps it's better to link these statically too for consistency.

File: plugins/newarc.ex/Framework/Src/farlng2.cpp
Comment (alabuzhev) [inline plugins/newarc.ex/Framework/Src/farlng2.cpp:1]: This plugin is a deep legacy and is not part of the build, it's ok to ignore it.

File: plugins/proclist/Pclass.cpp [hunk 2]
```diff
@@ -30,7 +30,11 @@ class StrTok
 		tok(token)
 	{
 		buf.reset(wcsdup(str));
+#if defined(_UCRT)
```
Comment (alabuzhev) [inline plugins/proclist/Pclass.cpp:33]: Is there any reason to not use `std::wcstok(buf.get(), token, &pt);` unconditionally?

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -6,6 +6,10 @@ on:
   pull_request:
     branches: [ master ]
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
```
Comment (alabuzhev) [inline .github/workflows/build.yml:11]: Nice.

**Comment by bitraid** [comment]
Did I somehow break appveyor?

**Review by alabuzhev — COMMENTED** [review]
Nice, thank you.

I just hope that will all these new build configurations we are still within github limits, whatever they are.

**Comment by alabuzhev** [comment]
> Did I somehow break appveyor?

Looking at the generated artefacts, all the hlf files are damaged.
Likely related to "Update gawk, m4 binaries" or "mkhlf.awk to utf-8".

**Comment by bitraid** [comment]
> I just hope that will all these new build configurations we are still within github limits, whatever they are.

Nah, we 're good. There are no limits for public projects, only for private ones.

**Comment by bitraid** [comment]
> > Did I somehow break appveyor?
> 
> Looking at the generated artefacts, all the hlf files are damaged. Likely related to "Update gawk, m4 binaries" or "mkhlf.awk to utf-8".

Ok, I'll revert the binaries first to see if it gets fixed.

File: far/exception_handler_test.cpp [hunk 4]
```diff
@@ -502,7 +502,11 @@ WARNING_POP()
 	static void seh_assertion_failure()
 	{
 		if ([[maybe_unused]] volatile const auto Condition = true)
+#if !defined(_MSC_VER) && (defined(_ARM_) || defined(_ARM64_))
+			DebugBreak();
```
[the remark this answers is not in the document]
Comment (bitraid) [reply far/exception_handler_test.cpp:506]: Yes, it's not defined. I'll remote it then.

File: far/farversion.m4 [hunk 5]
```diff
@@ -32,6 +32,8 @@ M4_MACRO_DEFINE(BUILD_PRERELEASE, m4_ifelse(
 M4_MACRO_DEFINE(BUILD_PLATFORM, m4_ifelse(
 	FARBIT, 64, x64,
 	FARBIT, 32, x86,
+	FARBIT, ARM64, arm64,
```
[the remark this answers is not in the document]
Comment (bitraid) [reply far/farversion.m4:35]: I was going to have it capitalized first, but then I saw some programs had it lower case. I'll change it.

File: plugins/proclist/Pclass.cpp (hunk 2 above)
Comment (bitraid) [reply plugins/proclist/Pclass.cpp:33]: I won't work for mingw-msvcrt:
https://github.com/mingw-w64/mingw-w64/blob/e0d90587c9825614fe59eda9066ad12fe72ff9bf/mingw-w64-headers/crt/string.h#L147-L151

File: plugins/luamacro/luasdk/ARM32/lua51.dll
Comment (bitraid) [reply plugins/luamacro/luasdk/ARM32/lua51.dll:1]: It's linked against ucrt, which is part of Windows 10 and later (where arm is supported). This happens for all other binaries of both arm32/arm64 from llvm-mingw (in contrast win vc). Are you sure?

Comment (alabuzhev) [reply plugins/luamacro/luasdk/ARM32/lua51.dll:1]: Not absolutely sure - I've seen quite a few issues with ucrt on older OSes, but not on 10, so I'm not saying it won't work.
My only argument is consistency: our prod binaries are vc, what happens with llvm-mingw is not that important.

---
By the way, do we need to "officially" publish ARM32 builds?
I think everything should work already, but no one ever asked.

Comment (bitraid) [reply plugins/luamacro/luasdk/ARM32/lua51.dll:1]: That makes sense. Even more so if you're thinking of providing an arm32 release.

**Comment by alabuzhev** [comment]
Looks like all is good now.

Thanks, merged.

**Comment by bitraid** [comment]
Awesome. Thanks.

**Pull request merged**